tapdb/sqlc/migrations/000024_universe_optimization_indexes_queries.up.sql

```diff
@@ -0,0 +1,53 @@
+-- Core universe_roots indexes
+CREATE INDEX IF NOT EXISTS idx_universe_roots_namespace 
+ON universe_roots(namespace_root, id, asset_id, group_key, proof_type);
+
+CREATE INDEX IF NOT EXISTS idx_universe_roots_issuance 
+ON universe_roots(proof_type, group_key, id, asset_id, namespace_root);
+
+-- Universe leaves optimization
+CREATE INDEX IF NOT EXISTS idx_universe_leaves_lookup 
+ON universe_leaves(
+    leaf_node_namespace, minting_point, script_key_bytes,
+    id, leaf_node_key, universe_root_id, asset_genesis_id
+);
+
+CREATE INDEX IF NOT EXISTS idx_universe_leaves_sort 
+ON universe_leaves(leaf_node_namespace, id, minting_point, script_key_bytes);
+
+-- MSSMT nodes optimization
+CREATE INDEX IF NOT EXISTS idx_mssmt_nodes_namespace 
+ON mssmt_nodes(namespace, hash_key, key, value, sum);
```

Comment:
We do we include the value and sum in the index? AFAIK we never use those in an actual query, or do we?

Reply:
Did some digging and it looks like we `SELECT` either `sum` or `value`:

`value`:
- Used in the `QueryUniverseLeaves` query, selected as `genesis_proof`.

`sum`:
- Used in the `FetchUniverseRoot` query, selected as `root_sum`.
- Used in the `QueryUniverseLeaves` query, selected as `sum_amt`.
- Used in the `QueryUniverseAssetStats` query in an aggregation to calculate the total supply.

My understanding is using this `covered` index helps improve these queries by not accessing the main tables:
https://www.postgresql.org/docs/current/indexes-index-only-scans.html

Reply:
Sorry, I was a bit imprecise. I mean we never use those values in the `WHERE` or `GROUP BY` clauses of a query, so an index shouldn't really be crucial.
Or is the idea to have a [covering index](https://www.postgresql.org/docs/current/indexes-index-only-scans.html) so we don't have to do a data area lookup for these queries at all?

Reply:
Yes, I _think_ here that with a covering index we could bypass table scans at all. I think though that it might not be worth the storage overhead, so I went ahead and made this a minimal index instead.